subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/RoleBasedConfigurationContainerInternal.java

```diff
@@ -0,0 +1,257 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+import org.gradle.api.Action;
+import org.gradle.api.GradleException;
+import org.gradle.internal.Actions;
+
+/**
+ * Extends {@link ConfigurationContainerInternal} with methods that can use {@link ConfigurationRole}s to
+ * define the allowed usage of a configuration at creation time.
+ */
+public interface RoleBasedConfigurationContainerInternal extends ConfigurationContainerInternal {
+    /**
+     * Creates a new configuration in the same manner as {@link #createWithRole(String, ConfigurationRole, boolean)}
+     * using the role of {@link ConfigurationRoles#INTENDED_CONSUMABLE}.
+     */
+    default ConfigurationInternal consumable(String name, boolean lockRole) {
+        return createWithRole(name, ConfigurationRoles.INTENDED_CONSUMABLE, lockRole);
+    }
+
+    /**
+     * Creates a new configuration in the same manner as {@link #createWithRole(String, ConfigurationRole, boolean)}
+     * using the role of {@link ConfigurationRoles#INTENDED_RESOLVABLE}.
+     */
+    default ConfigurationInternal resolvable(String name, boolean lockRole) {
+        return createWithRole(name, ConfigurationRoles.INTENDED_RESOLVABLE, lockRole);
+    }
+
+    /**
+     * Creates a new configuration in the same manner as {@link #createWithRole(String, ConfigurationRole, boolean)}
+     * using the role of {@link ConfigurationRoles#INTENDED_RESOLVABLE_BUCKET}.
+     */
+    @SuppressWarnings("deprecation")
+    default ConfigurationInternal resolvableBucket(String name, boolean lockRole) {
+        return createWithRole(name, ConfigurationRoles.INTENDED_RESOLVABLE_BUCKET, lockRole);
+    }
+
+    /**
+     * Creates a new configuration in the same manner as {@link #createWithRole(String, ConfigurationRole, boolean)}
+     * using the role of {@link ConfigurationRoles#INTENDED_BUCKET}.
+     */
+    default ConfigurationInternal bucket(String name, boolean lockRole) {
+        return createWithRole(name, ConfigurationRoles.INTENDED_BUCKET, lockRole);
+    }
+
+    /**
+     * Creates a new configuration in the same manner as {@link #createWithRole(String, ConfigurationRole, boolean)}
+     * using the role of {@link ConfigurationRoles#INTENDED_CONSUMABLE} that is <strong>NOT</strong> locked
+     * against further usage mutations.
+     */
+    default ConfigurationInternal consumable(String name) {
+        return consumable(name, false);
```

Comment:
When we make some of these methods public, we will want the builder methods that accept no boolean for lockRole to be locked by default (and based on our conversation we will not be making the boolean-accepting methods public)

We can either make these default to true now, or defer that to another PR but these methods will need to eventually pass in true.

Reply:
Per discussion, I want to leave this until we actually make some of these role create methods public.  At that point we should flip the default to true and adjust.